pkg/operator/defragcontroller/defragcontroller.go

```diff
@@ -0,0 +1,188 @@
+	timeoutDuration                 = 30 * time.Second
+)
+
+// DefragController observes the operand state file for fragmentation
+type DefragController struct {
+	operatorClient       v1helpers.OperatorClient
+	etcdClient           etcdcli.EtcdClient
+	infrastructureLister configv1listers.InfrastructureLister
+}
+
+func NewDefragController(
+	operatorClient v1helpers.OperatorClient,
+	etcdClient etcdcli.EtcdClient,
+	infrastructureLister configv1listers.InfrastructureLister,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &DefragController{
+		operatorClient:       operatorClient,
+		etcdClient:           etcdClient,
+		infrastructureLister: infrastructureLister,
+	}
+	return factory.New().ResyncEvery(10*time.Minute).WithInformers(
+		operatorClient.Informer(),
+	).WithSync(c.sync).ToController("DefragController", eventRecorder.WithComponentSuffix("defrag-controller"))
+}
+
+func (c *DefragController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	err := c.checkDefrag(ctx, syncCtx.Recorder())
+	if err != nil && !errors.Is(err, context.Canceled) {
+		_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "DefragControllerDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "Error",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			syncCtx.Recorder().Warning("DefragControllerUpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "DefragControllerDegraded",
+			Status: operatorv1.ConditionFalse,
+			Reason: "AsExpected",
+		}))
+	return updateErr
+}
+
+func (c *DefragController) checkDefrag(ctx context.Context, recorder events.Recorder) error {
+	controlPlaneTopology, err := ceohelpers.GetControlPlaneTopology(c.infrastructureLister)
+	if err != nil {
+		return err
+	}
+	// defrag is blocking and can only be safely performed in HighlyAvailableTopologyMode
+	if controlPlaneTopology != configv1.HighlyAvailableTopologyMode {
+		klog.V(4).Infof("DefragController is disabled for control plane topology: %s", controlPlaneTopology)
+		return nil
+	}
+
+	etcdMembers, err := c.etcdClient.MemberList()
+	if err != nil {
+		return err
+	}
+
+	// dont defrag if any of the cluster members are unhealthy
+	memberHealth, err := c.etcdClient.MemberHealth()
+	if err != nil {
+		return err
+	}
+	if !etcdcli.IsClusterHealthy(memberHealth) {
+		return fmt.Errorf("cluster is unhealthy: %s", memberHealth.Status())
+	}
+
+	var endpointStatus []*clientv3.StatusResponse
+	var leader *clientv3.StatusResponse
+	for _, member := range etcdMembers {
+		status, err := c.etcdClient.Status(ctx, member)
+		if err != nil {
+			return err
+		}
+		if leader == nil && status.Leader == member.ID {
+			leader = status
+			continue
+		}
+		endpointStatus = append(endpointStatus, status)
+	}
+
+	// leader last if possible
```

Comment:
I guess this is on best effort basis, for example if by chance there is a re-election by the time we finish these checks, we would not do this correct?

Reply:
right the checks are very fast but its possible but not probable

Reply:
Could we add a log here then, just to see if this is the case (might be good for troubleshooting), e.g. log level debug "Appending leader last".

Reply:
sure